File: scripts/LevelCurves.m
function LevelCurves(x_init, y_init, start_direction, dx, dy, L, N, onlySquare)
%
% Method of perpendicular: 
% see figure that doesn't exist
%
%
% Square construction:
% Name convention for for vertices and sides. The starting direction is 
% perpendicular to the sides and pointing outside the square (e.g.
% start_direction=3 means --->, start_direction=1 means <---).
% 
%                 2
%         2---------------3
%         |               | 
%         |               |
%       1 |               | 3
%         |               |
%         |               |
%         1---------------4
%                 4
%
%
tol         = 1e-4;
DEBUG       = 0;
ShowTestFig = 1;

xmin = -1;
xmax =  1;
nx   = 500;
ymin = -1;
ymax =  1;
ny   = 500;

x1d = linspace(xmin,xmax,nx);
y1d = linspace(ymin,ymax,ny);

[x,y] = meshgrid(x1d,y1d);
z     = func(x,y);

if DEBUG
    dx = 0.08;
    dy = 0.08;

    xp_vec = [-0.4949, -0.798 , 0.4949,  0.2323];
    yp_vec = [-0.2323,  0.6768, 0.7172, -0.3535];
    x0_vec = [-0.3131, -0.6768, 0.5758,  0.3939];
    y0_vec = [-0.3737,  0.7374, 0.6162, -0.2121];

    for i=1:length(x0_vec)
        x0 = x0_vec(i);
        y0 = y0_vec(i);
        xp = xp_vec(i);
        yp = yp_vec(i);

        z0 = func(x0, y0);

        square = CreateSquare(x0, y0, dx, dy, xmin, xmax, ymin, ymax);
        xs = square(:,1);
        ys = square(:,2);
        x1 = square(1,1);
        x2 = square(4,1); 
        y1 = square(1,2);
        y2 = square(2,2);
        side_number = FindSquareSide(z0, x0, y0, square, xp, yp);
        if side_number==1 || side_number==3
            a         = y1;
            b         = y2;
            isyaxis   = 1;
        else
            a         = x1;
            b         = x2; 
            isyaxis   = 0;
        end
        figure
        contour(x,y,z, 50)
        hold on 
        scatter(xs,ys, 'x', 'MarkerEdgeColor', 'r')
        scatter(x0,y0, 'o', 'filled', 'MarkerFaceColor', 'g')
        scatter(xp,yp, 'o', 'filled', 'MarkerFaceColor', 'r')
        
        [xc,yc,xd,yd] = FindPointsForRootFinder(xp,yp,x0,y0,0.1);
        xb = 2*x0-xp;
        yb = 2*y0-yp;
        scatter(xc,yc, 'filled', 'MarkerFaceColor', 'b')
        scatter(xd,yd, 'filled', 'MarkerFaceColor', 'b')
        scatter(xb,yb, 'filled', 'MarkerFaceColor', 'c')
        plot([xp,xb], [yp,yb], 'r')
        plot([xc,xd], [yc,yd], 'b')
        pbaspect([1 1 1])
        disp('square:')
        disp(square)
        fprintf('a           : %f\n', a)
        fprintf('b           : %f\n', b)
        fprintf('isyaxis     : %d\n', isyaxis)
        fprintf('side_number : %d\n', side_number)
        disp('---------------------------------------------------')
        disp('Press Enter to go to the next point...')
        pause
    end
    return
end

tic 

xp = x_init;
yp = y_init;
M  = func(xp,yp);
root_function = @(x,y) (func(x,y)-M);

fprintf('Searching level-curve for M=%.3f starting from (x,y)=(%.3f,%.3f)\n', M, xp, yp)

% find second point 
switch start_direction
    case 1
        a       = yp-dy;
        b       = yp+dy;
        c       = xp-dx;
        isyaxis = 1;
    case 2
        a       = xp-dx;
        b       = xp+dx;
        c       = yp+dy;
        isyaxis = 0;
    case 3
        a       = yp-dy;
        b       = yp+dy;
        c       = xp+dx;
        isyaxis = 1;
    case 4
        a       = xp-dx;
        b       = xp+dx;
        c       = yp-dy;
        isyaxis = 0;
    otherwise
        error('start_direction must be 1,2,3 or 4')
end
if isyaxis
    [xsc, ysc] = BisectionAlongLine(c, a, c, b, tol, root_function);
else
    [xsc, ysc] = BisectionAlongLine(a, c, b, c, tol, root_function);
end

points      = zeros(N+2,2);
points(1,:) = [xp, yp ];
points(2,:) = [xsc,ysc];

% squares method
if ShowTestFig
    figure
    contour(x,y,z, 50)
    hold on 
    scatter(xp,  yp, 80, 'o', 'filled', 'MarkerFaceColor', 'm')
    scatter(xsc,ysc, 80, 'x', 'm', 'LineWidth', 2 )
end

iter = 0;
for i=1:N
    
    if ShowTestFig
        xb = 2*xsc-xp;
        yb = 2*ysc-yp;
        xp_old = xp;
        yp_old = yp;
        xsc_old = xsc;
        ysc_old = ysc;
    end
    
    if ~onlySquare
        [x1_bis,y1_bis,x2_bis,y2_bis] = FindPointsForRootFinder(xp,yp,xsc,ysc,L);
        [xsc_tmp,ysc_tmp,iter] = BisectionAlongLine(x1_bis, y1_bis, x2_bis, y2_bis, tol, root_function);
        %fprintf('%d iterations in BisectionAlongLine\n', iter)
    end
    if iter>0
        xp  = xsc;
        yp  = ysc;
        xsc = xsc_tmp;
        ysc = ysc_tmp;
    else
        square = CreateSquare(xsc, ysc, dx, dy, xmin, xmax, ymin, ymax);
        side_number = FindSquareSide(M, xsc, ysc, square, xp, yp);
        x1 = square(1,1);
        x2 = square(4,1); 
        y1 = square(1,2);
        y2 = square(2,2);
        switch side_number
            case 1
                a       = y1;
                b       = y2;
                c       = x1;
                isyaxis = 1;
            case 2
                a       = x1;
                b       = x2;
                c       = y2;
                isyaxis = 0;
            case 3
                a       = y1;
                b       = y2;
                c       = x2;
                isyaxis = 1;
            case 4
                a       = x1;
                b       = x2;
                c       = y1;
                isyaxis = 0;
            otherwise 
                error('Side of the square not found! Try to change the initial direction')
        end

        xp = xsc;
        yp = ysc;

        if isyaxis
            [xsc,ysc] = BisectionAlongLine(c, a, c, b, tol, root_function);
        else
            [xsc,ysc] = BisectionAlongLine(a, c, b, c, tol, root_function);
        end
    end
    
    points(i+2,:) = [xsc, ysc];
    
    if ShowTestFig
        scatter(xsc,ysc, 80, 'x', 'g', 'LineWidth', 2 )
        if iter==0
            scatter(xsc_old,ysc_old, 70, 'o', 'MarkerEdgeColor', 'k')
            scatter(square(:,1), square(:,2), 'x', 'MarkerEdgeColor', 'r')
        else
            scatter(x1_bis,y1_bis, 'filled', 'MarkerFaceColor', 'b')
            scatter(x2_bis,y2_bis, 'filled', 'MarkerFaceColor', 'b')
            scatter(xb,yb, 'filled', 'MarkerFaceColor', 'c')
            plot([xp_old,xb], [yp_old,yb], 'r')
            plot([x1_bis,x2_bis], [y1_bis,y2_bis], 'b')
        end
        xlim([xmin xmax])
        ylim([ymin ymax])
        drawnow
    end
    
    if tooclose(xsc, xmin)   || tooclose(xsc, xmax) || ...
       tooclose(ysc, ymax)   || tooclose(ysc, ymin) || ...
       (abs(xsc-x_init)<dx && abs(ysc-y_init)<dy)
        fprintf('Stopped at %d iteration\n', i)
        break
    end
end

toc

return 

function bool = tooclose(a,b)
eps = 1e-3;
if abs(a-b)<eps
    bool = 1;
else
    bool = 0;
end
return

function z = func(x,y)
z = sin(4*x).*cos(4*y);
%z = sin(x.*y);
return

function [xroot, yroot, iter] = BisectionAlongLine(x1, y1, x2, y2, tol, myfunc)
verbose = 0;
itermax = 100;
tol2    = tol*tol; % avoid sqrt() in if-statement

f1 = myfunc(x1,y1);
f2 = myfunc(x2,y2);

iter    = 0;
if f1*f2>0
    if verbose
        fprintf('root not found! Returning (xroot,yroot)=(%f,%f)\n',x1,y1)
    end
    xroot = x1;
    yroot = y1;
    return
end

while 1
    iter = iter + 1;
    
    xm = (x1+x2)/2;
    ym = (y1+y2)/2;
    fm = myfunc(xm,ym);
    
    if ((x1-x2)^2+(y1-y2)^2)<tol2 || iter>itermax
        if iter>itermax
            warning('reached itermax=%d!', itermax)
        end
        xroot = xm;
        yroot = ym;
        if verbose
            fprintf('tol=%e\titer=%3d\t(%7.4f,%7.4f)\n',tol,iter,xm,ym)
        end
        return
    end
    
    if f1*fm<0
        x2 = xm;
        y2 = ym;
    else
        if f1*fm>0
            x1 = xm;
            y1 = ym;
            f1 = fm;
        else
            xroot = xm;
            yroot = ym;
            return
        end
    end
end
return

function points = CreateSquare(x0, y0, dx, dy, xmin, xmax, ymin, ymax)
eps = 1e-10;
x1  = x0-dx;
x2  = x0+dx;
y1  = y0-dy;
y2  = y0+dy;
if x1<xmin
    x1 = xmin+eps;
end
if x2>xmax
    x2 = xmax-eps;
end
if y1<ymin
    y1 = ymin+eps;
end
if y2>ymax
    y2 = ymax-eps;
end
points      = zeros(4,2);
points(1,:) = [x1, y1];
points(2,:) = [x1, y2];
points(3,:) = [x2, y2];
points(4,:) = [x2, y1];
return

function [side_number, logic_vec] = FindSquareSide(k, x0, y0, square, x_previous, y_previous)
k1 = func(square(1,1), square(1,2));
k2 = func(square(2,1), square(2,2));
k3 = func(square(3,1), square(3,2));
k4 = func(square(4,1), square(4,2));

intervals      = zeros(4,2);
intervals(1,:) = sort([k1,k2]);
intervals(2,:) = sort([k2,k3]);
intervals(3,:) = sort([k3,k4]);
intervals(4,:) = sort([k4,k1]);
logic_vec      = zeros(4,1);
for i=1:4
    if k>intervals(i,1) && k<intervals(i,2)
        logic_vec(i) = 1;
    end
end

if sum(logic_vec)==0
    side_number = 0;
end

if logic_vec(1) && logic_vec(3)
    if x0>x_previous
        side_number = 3;
    else
        side_number = 1;
    end
elseif logic_vec(2) && logic_vec(4)
    if y0>y_previous
        side_number = 2;
    else
        side_number = 4;
    end
elseif logic_vec(1) && logic_vec(2)
    if y0>y_previous
        side_number = 2;
    else
        side_number = 1;
    end
elseif logic_vec(2) && logic_vec(3)
    if y0>y_previous
        side_number = 2;
    else
        side_number = 3;
    end
elseif logic_vec(3) && logic_vec(4)
    if y0>y_previous
        side_number = 3;
    else
        side_number = 4;
    end
elseif logic_vec(1) && logic_vec(4)
    if y0>y_previous
        side_number = 1;
    else
        side_number = 4;
    end
end
return

function [x1,y1,x2,y2] = FindPointsForRootFinder(xp,yp,x0,y0,L)
if xp==x0
    % vertical line
    x1 = x0;
    y1 = y0-L;
    x2 = x0;
    y2 = y0+L;
else
    % angular coeff of the line passing for (xp,yp) and (x0,y0)
    m = (y0-yp)/(x0-xp);
    %q = y0-m*x0;

    % middle point of (x1,y1) and (x2,y2)
    xm = 2*x0-xp;
    ym = 2*y0-yp;

    % line perpendicular to y=m*x and passing for the middle point
    M = -1/m;
    Q = ym - M*xm;
    
    % find (x1,y1) and (x2,y2)
    a  = (1+M^2);
    b  = 2*M*Q-2*xm-2*ym*M;
    c  = xm^2+ym^2+Q^2-2*ym*Q-L^2;
    x1 = (-b-sqrt(b^2-4*a*c))/(2*a);
    y1 = M*x1+Q;
    x2 = (-b+sqrt(b^2-4*a*c))/(2*a);
    y2 = M*x2+Q;
end
return

File: scripts/LevelCurves.cpp
#include "LevelCurves.h"
/*
TODO
- add comments 
- fix notation: x1/x2 used many times for different things
*/

double L2SquareDistance(double x1, double y1, double x2, double y2){
   return (x1-x2)*(x1-x2)+(y1-y2)*(y1-y2); 
}
int BisectionForLevelCurves(double x1, double y1, double x2, double y2, double (*func)(double,double), double K, double tol, 
              double &xroot, double &yroot){
    double tol2;
    int itermax, iter;
    double f1, f2;
    double xm, ym, fm;

    tol2    = tol*tol;
    itermax = 100;
    iter    = -1;

    f1 = func(x1,y1)-K;
    f2 = func(x2,y2)-K;

    if (f1*f2>0){
        return iter;
    }
    
    iter++;
    
    while (1) {
        xm = (x1+x2)/2;
        ym = (y1+y2)/2;

        if (L2SquareDistance(x1,y1,x2,y2)<tol2 || iter>itermax){
            if (iter>itermax) {
                printf("reached itermax=%d!\n", itermax);
            }
            xroot = xm;
            yroot = ym;
            return iter;
        }
        
        fm = func(xm,ym)-K;
        iter++;
        
        if (f1*fm<0){
            x2 = xm;
            y2 = ym;
        } else if (f1*fm>0){
            x1 = xm;
            y1 = ym;
            f1 = fm;
        } else {
            xroot = xm;
            yroot = ym;
            return iter;
        }
    }
}

void CreateSquare(double x0, double y0, double ds, double xmin, double xmax, double ymin, double ymax, 
                  double square[][2]){
    double eps = 1e-10;
    double x1, x2, y1, y2;
    x1  = x0-ds;
    x2  = x0+ds;
    y1  = y0-ds;
    y2  = y0+ds;
    if (x1<xmin){
        x1 = xmin+eps;
    }
    if (x2>xmax){
        x2 = xmax-eps;
    }
    if (y1<ymin){
        y1 = ymin+eps;
    }
    if (y2>ymax){
        y2 = ymax-eps;
    }
    square[0][0] = x1;
    square[0][1] = y1;
    square[1][0] = x1;
    square[1][1] = y2;
    square[2][0] = x2;
    square[2][1] = y2;
    square[3][0] = x2;
    square[3][1] = y1;
    return;
}

int FindSquareSide(double K, double x0, double y0, double square[][2], double x_previous, double y_previous, 
                   double (*func)(double, double)){
    double k1, k2, k3, k4;
    double intervals[4][2]={{0}};
    double logic_vec[4]={0};

    k1 = func(square[0][0], square[0][1]);
    k2 = func(square[1][0], square[1][1]);
    k3 = func(square[2][0], square[2][1]);
    k4 = func(square[3][0], square[3][1]);

    if (k1<k2){
        intervals[0][0] = k1;
        intervals[0][1] = k2;
    } else {
        intervals[0][0] = k2;
        intervals[0][1] = k1;
    }

    if (k2<k3){
        intervals[1][0] = k2;
        intervals[1][1] = k3;
    } else {
        intervals[1][0] = k3;
        intervals[1][1] = k2;
    }

    if (k3<k4){
        intervals[2][0] = k3;
        intervals[2][1] = k4;
    } else {
        intervals[2][0] = k4;
        intervals[2][1] = k3;
    }

    if (k4<k1){
        intervals[3][0] = k4;
        intervals[3][1] = k1;
    } else {
        intervals[3][0] = k1;
        intervals[3][1] = k4;
    }


    for(int i=0; i<4; i++){
        if (K>intervals[i][0] && K<intervals[i][1])
            logic_vec[i] = 1;
    }

    int side_number;

    if ((logic_vec[0]+logic_vec[1]+logic_vec[2]+logic_vec[3])<1e-12)
        side_number = 0;

    if (logic_vec[0] && logic_vec[2]){
        if (x0>x_previous){
            side_number = 3;
        } else {
            side_number = 1;
        }
    } else if (logic_vec[1] && logic_vec[3]) {
        if (y0>y_previous){
            side_number = 2;
        } else {
            side_number = 4;
        }
    } else if (logic_vec[0] && logic_vec[1]){
        if (y0>y_previous){ 
            side_number = 2;
        } else {
            side_number = 1;
        }
    } else if (logic_vec[1] && logic_vec[2]){
        if (y0>y_previous){
            side_number = 2;
        } else {
            side_number = 3;
        }
    } else if (logic_vec[2] && logic_vec[3]){
        if (y0>y_previous){
            side_number = 3;
        } else {
            side_number = 4;
        }
    } else if (logic_vec[0] && logic_vec[3]){
        if (y0>y_previous){
            side_number = 1;
        } else {
            side_number = 4;
        }
    }
    return side_number;
}

void FindPointsForRootFinder(double xp,  double yp,  double x0,  double y0, double L, 
                             double &x1, double &y1, double &x2, double &y2){
    if (xp==x0){
        x1 = x0;
        y1 = y0-L;
        x2 = x0;
        y2 = y0+L;
    } else {
        double m, xm, ym, M, Q, a, b, c;
        // angular coeff of the line passing for (xp,yp) and (x0,y0)
        m = (y0-yp)/(x0-xp);
        // middle point of (x1,y1) and (x2,y2)
        xm = 2*x0-xp;
        ym = 2*y0-yp;
        // line perpendicular to y=m*x and passing for the middle point
        M = -1/m;
        Q = ym - M*xm;
        // find (x1,y1) and (x2,y2)
        a  = (1+M*M);
        b  = 2*M*Q-2*xm-2*ym*M;
        c  = xm*xm+ym*ym+Q*Q-2*ym*Q-L*L;
        x1 = (-b-sqrt(b*b-4*a*c))/(2*a);
        y1 = M*x1+Q;
        x2 = (-b+sqrt(b*b-4*a*c))/(2*a);
        y2 = M*x2+Q;
    }

    return;
}

bool FindSecondPoint(double xp, double yp, double ds, int start_direction, 
                     double (*func)(double,double), double K, double tol, double &x0, double &y0){
    printf("Searching level-curve for M=%.3f starting from (x,y)=(%.3f,%.3f)\n", K, xp, yp);
    double a,b,c;
    bool isyaxis;
    // find second point 
    if (start_direction==1) {
        a       = yp-ds;
        b       = yp+ds;
        c       = xp-ds;
        isyaxis = 1;
    } else if (start_direction==2){
        a       = xp-ds;
        b       = xp+ds;
        c       = yp+ds;
        isyaxis = 0;
    } else if (start_direction==3){
        a       = yp-ds;
        b       = yp+ds;
        c       = xp+ds;
        isyaxis = 1;
    } else if (start_direction==4){
        a       = xp-ds;
        b       = xp+ds;
        c       = yp-ds;
        isyaxis = 0;
    } else {
        printf("start_direction must be 1,2,3 or 4");
        return 0;
    }

    if (isyaxis) {
        return BisectionForLevelCurves(c, a, c, b, func, K, tol, x0, y0);
    } else {
        return BisectionForLevelCurves(a, c, b, c, func, K, tol, x0, y0);
    }
}

int FindPoints(double x_init, double y_init, double ds, double x0, double y0,
                    double xmin, double xmax, double ymin, double ymax, double L, double tol, double **points, int N,
                    double (*func)(double,double), double K){
    double x1_bis, y1_bis, x2_bis, y2_bis;
    double x0_tmp, y0_tmp;

    // variables needed for the square
    double square[4][2];
    double x1,y1,x2,y2;
    int side_number;
    double a,b,c;
    bool isyaxis;

    double xp, yp;
    xp = x_init;
    yp = y_init;
   
    // Find points 
    int iter_bisec;
    int points_counter = 2;
    int last_point_on_boundary = 0;
    
    // variables used for checking that the points for bisection are into the boundary
    int boundary_iters, boundary_iters_max=9, boundary_iters_max_reached;
    double L_tmp, puppet;
    
    for(int i=2; i<N; i++){
        if (!last_point_on_boundary){
            FindPointsForRootFinder(xp,yp,x0,y0,L,x1_bis,y1_bis,x2_bis,y2_bis);
           
            boundary_iters_max_reached = 0;
            
            // check if (x1_bis, y1_bis) is inside the boundary, otherwise change it
            L_tmp = L;
            boundary_iters = 0;
            while (x1_bis<xmin || x1_bis>xmax || y1_bis<ymin || y1_bis>ymax){
                L_tmp = L_tmp-L/10;
                FindPointsForRootFinder(xp,yp,x0,y0,L_tmp,x1_bis,y1_bis,puppet,puppet);
                boundary_iters++;
                if (boundary_iters>boundary_iters_max){
                    boundary_iters_max_reached = 1;
                    break;
                }
            }
            
            // check if (x2_bis, y2_bis) is inside the boundary, otherwise change it
            L_tmp = L;
            boundary_iters = 0;
            while (x2_bis<xmin || x2_bis>xmax || y2_bis<ymin || y2_bis>ymax){
                L_tmp = L_tmp-L/10;
                FindPointsForRootFinder(xp,yp,x0,y0,L_tmp,puppet,puppet,x2_bis,y2_bis);
                boundary_iters++;
                if (boundary_iters>boundary_iters_max){
                    boundary_iters_max_reached = 1;
                    break;
                }
            }
           
            if (!boundary_iters_max_reached) {
                // just a check
                if (x2_bis<xmin || x2_bis>xmax || y2_bis<ymin || y2_bis>ymax ||
                    x1_bis<xmin || x1_bis>xmax || y1_bis<ymin || y1_bis>ymax){
                    printf("+++ Warning! Initial points for bisection outside of boundary at point #%d ++\n", i); 
                }
                // bisection
                iter_bisec = BisectionForLevelCurves(x1_bis, y1_bis, x2_bis, y2_bis, func, K, tol, x0_tmp, y0_tmp);
            } else {
                iter_bisec = -1;
            }
        } else {
            iter_bisec = -1;
        }
        if (iter_bisec>=0){
            xp  = x0;
            yp  = y0;
            x0 = x0_tmp;
            y0 = y0_tmp;
        } else {
            printf("Using squares-method at iteration %d, (xp,yp)=(%9f,%9f)\n", i, xp, yp);
            CreateSquare(x0,y0,ds,xmin,xmax,ymin,ymax,square);
            x1 = square[0][0];
            x2 = square[3][0]; 
            y1 = square[0][1];
            y2 = square[1][1];
            side_number = FindSquareSide(K, x0, y0, square, xp, yp, func);
            if (side_number==1){
                a       = y1;
                b       = y2;
                c       = x1;
                isyaxis = 1;
            } else if (side_number==2) {
                a       = x1;
                b       = x2;
                c       = y2;
                isyaxis = 0;
            } else if (side_number==3) {
                a       = y1;
                b       = y2;
                c       = x2;
                isyaxis = 1;
            } else if (side_number==4) {
                a       = x1;
                b       = x2;
                c       = y1;
                isyaxis = 0;
            } else {
                printf("Side of the square not found! Try to change the initial direction\n");
            }

            xp = x0;
            yp = y0;

            if (isyaxis) {
                BisectionForLevelCurves(c, a, c, b, func, K, tol, x0, y0);
            } else {
                BisectionForLevelCurves(a, c, b, c, func, K, tol, x0, y0);
            }
        }
        
        points[i][0] = x0;
        points[i][1] = y0;
        points_counter++;

        if (x0<xmin || x0>xmax || y0<ymin || y0>ymax || 
            last_point_on_boundary || (fabs(x0-x_init)<=ds && fabs(y0-y_init)<=ds)){
            printf("Stopped at %d iteration\n", i);
            break;
        }
        
        if ((x0-xmin)<ds || (xmax-x0)<ds || (ymax-y0)<ds || (y0-ymin)<ds)
            last_point_on_boundary++;
    }
    return points_counter; 
}

void FindLevelCurve(double x_init, double y_init, double ds, int start_direction, double L, double tol, double
                   (*func)(double, double), double xmin, double xmax, double ymin, double ymax, 
                   char *fname, int maxpoints){
    
    double x0, y0,K;
    K = func(x_init,y_init);
    FindSecondPoint(x_init, y_init, ds, start_direction, func, K, tol, x0, y0);
    
    // allocate points
    double **points; 
    points = new double*[maxpoints]; 
    for (int i=0; i<maxpoints; i++)
        points[i] = new double[2];
    
    points[0][0] = x_init;
    points[0][1] = y_init;
    points[1][0] = x0;
    points[1][1] = y0;
    
    int iter;
    iter = FindPoints(x_init, y_init, ds, x0, y0, xmin, xmax, ymin, ymax, L, tol, points, maxpoints, func, K);
    
    FILE *fp;
    fp=fopen(fname,"w");
    for(int i=0; i<iter; i++)
        fprintf(fp, "%23.15e %23.15e\n", points[i][0], points[i][1]);
    fclose(fp);

    cout<<"Ouput printed on '"<<fname<<"'"<<endl;
    
    // free memory
    for(int i=0 ;i<maxpoints; i++)
        delete[] points[i]; 
    delete[] points;
}

